README.md
# OINP-updates

Send an email notification when there were any updates from OINP ([Ontario Immigrant Nominee Program](https://www.ontario.ca/page/2019-ontario-immigrant-nominee-program-updates))

| Example Email |
| ------------------------------------ |
| ![Example Eamil](example-email.png)  | 

Only support *Python3*.

## deploy

Simply deploy on the home directory of your server and serve by crontab.

Using [SendGrid](https://sendgrid.com/) to send email.

``` shell
# python3 pkgs
pip3 install beautifulsoup4 requests html5lib sendgrid

# deploy it in user's base directory
cd ~

# clone repo
git clone https://github.com/ottocho/OINP-updates

#
cd OINP-updates

#
mkdir db

# set config for email sending
vim ./config.py

    SG_API_KEY = 'YOUR_SEND_GRID_KEY'

    EMAIL_RECVERS = [
        'RECVER1@z.com',
        'RECVER2@z.com',
    ]
    EMAIL_DEVS = [
        'DEV@z.com',
    ]
    EMAIL_SENDER = 'SENDER@z.com'

# run
./main.py
```

## on cron

```
*/5 *  *   *   *    cd /home/otto_cho/OINP-updates/ && TZ='America/New_York' python3 main.py >>run.log 2>&1
```


CHANGELOG.md
# Changelog

## v5 - Mar. 07, 2019

- Refactor the code base

## v4 - Mar. 04, 2019

- Due to the OINP service migration from [ontarioimmigration.ca](http://www.ontarioimmigration.ca/en/pnp/OI_PNPNEW.html) to [onratio.ca](https://www.ontario.ca/page/2019-ontario-immigrant-nominee-program-updates), this service is updated to use new their new data source.

## v3 - Dec. 06, 2018

- Fix datetime parsing bug

## v2 - Dec. 25, 2017

- Send debug information

## v1 - Dec. 5, 2017

- fix the email config error

## v0 - Dec. 1, 2017

- first stable version


main.py
#!/usr/bin/env python3
#coding:utf8

"""
Author:       ottocho
Filename:     OINP-updates/main.py
Date:         2019-03-07 10:18 PM
Description:  fetch news and send email
"""

import os
import re
import json
import traceback
from datetime import datetime
from collections import defaultdict

import requests
from bs4 import BeautifulSoup

from mail import send_dev, send_msg
from config import *

def parse_date_title(l):
    '''
    use this line to be the starter of one news
    indicated line: `December 6th, 2018`
    '''
    if not l:
        return l
    PTN = r'''
    (?P<month>Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Sept|Oct|Nov|Dec
              |January|February|March|April|May|June|July|August
              |September|October|November|December)\.?[ ]+          # month
    (?P<day>[0-9]+)(?:st|nd|rd|th|ST|ND|RD|TH)?[ ]*,?[ ]*           # day
    (?P<year>20[21][67890])                                         # year
    '''
    rk = re.compile(PTN, re.VERBOSE)
    t = rk.findall(l)
    if len(t) == 0:
        return
    _dm, dd, dy = t[0]
    dm = M2D[_dm]
    return (int(dm), int(dd), int(dy))

def fetch_news_tags():
    '''
    get the main part of the news website
    '''
    headers = {
        'User-Agent': HTTP_USER_AGENT,
        'Accept': '*/*',
        'Content-Encoding': 'gzip',
    }
    value = requests.get(URL_DATA_SOURCE, timeout=HTTP_TIMEOUT, headers=headers).json()
    html = value['body']['und'][0]['safe_value']
    soup = BeautifulSoup(html, 'html5lib')
    return soup.body.children

def parse_news_tags():
    '''
    parse the content lines
    '''
    # get content first
    tags = fetch_news_tags()

    started = False
    new_trunk = []
    trunks = defaultdict(list)
    last_date = None
    for tag in tags:
        if tag.name == 'h2' and tag.has_attr('id') and tag['id'].startswith('section-'):
            # the Month tag
            continue
        if tag.string:
            # remove blank line
            tag_string = re.sub(r'\s', '', tag.string)
            if not tag_string:
                continue
        if tag.name == 'h3':
            new_date = parse_date_title(tag.string)
            if new_date: # begining of a new message
                if started:
                    trunks[last_date].append(new_trunk)
                started = True
                last_date = new_date
                new_trunk = [ str(tag) ]
                continue
        new_trunk.append(str(tag))

    result = []
    # make lines a plain list and give them index
    for date, newslist in trunks.items():
        dm, dd, dy = date
        sl = len(newslist)
        for idx, news in enumerate(newslist):
            idx = '%d-%02d-%02d-%02d' % (dy, dm, dd, (sl-idx))
            result.append((idx, news))
    return result

def get_last_update():
    if not os.path.isfile(PATH_LAST_UPDATE):
        return
    with open(PATH_LAST_UPDATE) as fobj:
        return json.loads(fobj.read())

def write_last_update(new_update):
    # update history data
    data = json.dumps(new_update, indent=4)
    with open(PATH_LAST_UPDATE, 'w') as fobj:
        fobj.write(data)
    with open(PATH_NEW_UPDATE, 'w') as fobj:
        fobj.write(data)

def main():
    print(datetime.now())

    # get new data from api
    new_update = parse_news_tags()
    # get last data from local cache data
    last_update = get_last_update()
    # write new data to local cache data
    write_last_update(new_update)

    if not last_update:
        print('initial')
        return

    # compare data
    keyset_new = set([_t[0] for _t in new_update])
    keyset_last = set([_t[0] for _t in last_update])
    keys_new = keyset_new - keyset_last

    # no update, return
    if not keys_new:
        print('no update')
        return

    # send notification!
    dict_new_update = dict(new_update)
    for kn in keys_new:
        print('update %s' % kn)
        nn = dict_new_update[kn]
        subject = 'UPDATE: Ontario Immigrant Nominee Program (%s)' % kn
        html = ''.join(nn)
        html = EMAIL_PREFIX + html
        send_msg(subject, html)

if __name__ == '__main__':
    try:
        main()

    except (IndexError, ConnectionResetError, requests.exceptions.Timeout, requests.exceptions.ContentDecodingError):
        print('time out')

        n = 1
        o = { 'n': n }
        # check time out time
        if os.path.isfile(PATH_TIMEOUT_REC):
            rb = open(PATH_TIMEOUT_REC, 'r+')
            o = json.loads(rb.read())
            n = o['n']
            rb.close()
        print('timeout #%d' % n)
        if n > HTTP_TIMEOUT_THRES:
            html = '<strong><pre>ERROR Timeout Detection</pre></strong>'
            html = EMAIL_PREFIX + html
            subject = 'ERROR: Ontario Immigrant Nominee Program'
            print(subject, html)
            send_dev(subject, html)
            n = 0

        # store timeout time
        wb = open(PATH_TIMEOUT_REC, 'w+')
        o['n'] = n + 1
        wb.write(json.dumps(o))
        wb.close()

    except Exception as e:
        exs = traceback.format_exc()
        html = '<strong><pre>%s</pre></strong>' % exs
        html = EMAIL_PREFIX + html
        subject = 'ERROR: Ontario Immigrant Nominee Program '
        print(subject, exs)
        send_dev(subject, html)


config.py
#!/usr/bin/env python3
#coding:utf8

"""
Author:       ottocho
Filename:     OINP-updates/config.py
Date:         2019-03-07 10:18 PM
Description:  configuration
"""

from datetime import datetime

YEAR = datetime.now().year

''' token for sendgrid '''
SG_API_KEY = ''

''' email settings '''
EMAIL_RECVERS = [
]
EMAIL_DEVS = [
]
EMAIL_SENDER = ''

''' data source URL '''
_HOST = 'api.ontario.ca'
_PATH = '/api/drupal/page%%2F%d-ontario-immigrant-nominee-program-updates' % YEAR
_GET_PARAMS = 'fields=nid,field_body_beta,body'
URL_DATA_SOURCE = 'https://%s/%s?%s' % (_HOST, _PATH, _GET_PARAMS)

''' URL posting OINP news '''
URL_NEWS = 'https://www.ontario.ca/page/%d-ontario-immigrant-nominee-program-updates' % YEAR
EMAIL_PREFIX = '<p><a target="_blank" href="%s">%s</a></p>' % (URL_NEWS, URL_NEWS)

''' path to file storing records '''
PATH_DB = 'db/'
PATH_LAST_UPDATE = PATH_DB + '/' + 'last.json'
PATH_NEW_UPDATE = PATH_DB + '/' + '%s.json' % datetime.now().strftime('%Y-%m%d-%H%M')
PATH_TIMEOUT_REC = PATH_DB + '/' + 'timeout.json'

''' http settings '''
HTTP_TIMEOUT = 15
HTTP_TIMEOUT_THRES = 500 # send email if timeout too much
HTTP_USER_AGENT = 'Mozilla/5.0 (Windows NT 5.1; rv:7.0.1) Gecko/20100101 Firefox/7.0.1'

''' month to day number '''
M2D = {
    "January": 1,
    "February": 2,
    "March": 3,
    "April": 4,
    "May": 5,
    "June": 6,
    "July": 7,
    "August": 8,
    "September": 9,
    "October": 10,
    "November": 11,
    "December": 12,
    "Jan": 1,
    "Feb": 2,
    "Mar": 3,
    "Apr": 4,
    "May": 5,
    "Jun": 6,
    "Jul": 7,
    "Aug": 8,
    "Sep": 9,
    "Sept": 9,
    "Oct": 10,
    "Nov": 11,
    "Dec": 12,
}


send_mail.py
#!/usr/bin/env python3
#coding:utf8

"""
Author:       ottocho
Filename:     send_mail.py
Date:         2017-11-14 21:10
Description:

send mail via sendgrid

"""

import os

import sendgrid
from sendgrid.helpers.mail import *

SG_KEY_FILE = 'private/SG_API_KEY'
SENDER_FILE = 'private/SENDER'
RECVER_FILE = 'private/RECVER'
DEBUGGER_FILE = 'private/DEBUGGER'

def get_email_lines(filename):
    '''
    get email list from file
    '''
    with open(filename) as fobj:
        rs = [ i.strip() for i in fobj.read().split('\n') ]
        rs = [ i for i in rs if i and not i.startswith('#') ]
        return rs

def send_mail(subject, html):
    # get config
    apikey = open(SG_KEY_FILE).read().strip()
    sender_email = open(SENDER_FILE).read().strip()
    rs = get_email_lines(RECVER_FILE)

    for recver_email in rs:
        # api send
        sg = sendgrid.SendGridAPIClient(apikey=apikey)
        from_email = Email(sender_email)
        to_email = Email(recver_email)
        content = Content("text/html", html)
        mail = Mail(from_email, subject, to_email, content)
        response = sg.client.mail.send.post(request_body=mail.get())
        print(recver_email)
        print(response.status_code)
        print(response.body)
        print(response.headers)

def send_debug(subject, html):
    # get config
    apikey = open(SG_KEY_FILE).read().strip()
    sender_email = open(SENDER_FILE).read().strip()
    recver_email = open(DEBUGGER_FILE).read().strip()

    # api send
    sg = sendgrid.SendGridAPIClient(apikey=apikey)
    from_email = Email(sender_email)
    to_email = Email(recver_email)
    content = Content("text/html", html)
    mail = Mail(from_email, subject, to_email, content)
    response = sg.client.mail.send.post(request_body=mail.get())
    print(recver_email)
    print(response.status_code)
    print(response.body)
    print(response.headers)

if __name__ == '__main__':
    subject = 'UPDATE! Ontario Immigrant Nominee Program'
    html = '<p><strong>test mail</strong></p>'
    send_mail(subject, html)
    # send_debug(subject, html)


mail.py
#!/usr/bin/env python3
#coding:utf8

"""
Author:       ottocho
Filename:     OINP-updates/mail.py
Date:         2019-03-07 10:18 PM
Description:  send mail via sendgrid
"""

import sendgrid
from sendgrid.helpers.mail import *

from config import SG_API_KEY, EMAIL_SENDER, EMAIL_RECVERS, EMAIL_DEVS
if not (SG_API_KEY and EMAIL_SENDER and EMAIL_RECVERS and EMAIL_DEVS):
    raise Exception('config error')

def _send(subject, html, receivers):
    for email_recver in receivers:
        sg = sendgrid.SendGridAPIClient(apikey=SG_API_KEY)
        from_email = Email(EMAIL_SENDER)
        to_email = Email(email_recver)
        content = Content("text/html", html)
        mail = Mail(from_email, subject, to_email, content)
        response = sg.client.mail.send.post(request_body=mail.get())
        print(email_recver)
        print(response.status_code)
        print(response.body)
        print(response.headers)

def send_msg(subject, html):
    return _send(subject, html, EMAIL_RECVERS)

def send_dev(subject, html):
    return _send(subject, html, EMAIL_DEVS)

if __name__ == '__main__':
    subject = 'UPDATE! Ontario Immigrant Nominee Program'
    html = '<p><strong>test mail</strong></p>'
    send_dev(subject, html)
